llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -7657,6 +7720,13 @@ static bool isGuaranteedNotToBeUndefOrPoison(
       }
       if (IsWellDefined)
         return true;
+
+      bool StartNeedsFreeze;
+      if (canFoldFreezeIntoRecurrence(
+              const_cast<PHINode *>(PN), const_cast<DominatorTree *>(DT),
```

Comment:
Is it possible to avoid the `const_cast` if you change the interface `canFoldFreezeIntoRecurrence` to be

```
Use *llvm::canFoldFreezeIntoRecurrence(
    const PHINode *PN, const DominatorTree *DT, bool &StartNeedsFreeze,
```
? I'm just a bit worried that callers of `isGuaranteedNotToBeUndefOrPoison` really expect the `PN` and `DT` objects not to change and by using this cast here you're violating some assumption.

Reply:
I tried marking DominatorTree as const and it seemed to work for me, although I agree it's hard to avoid the `const_cast` for the PHI node. However, I guess it's fine in this case because we're ignoring the result and we're not going to drop any flags. I see there are existing uses of const_cast elsewhere in ValueTracking.cpp too.

Can you update the interface to just mark the DominatorTree argument as const then?